### PR TITLE
Add FinishTrajectory() in SparsePosegraph

### PR DESCRIPTION
**Pull request by jihoonl**
Towards #560

### CODE REVIEW

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 1]
```diff
@@ -396,6 +401,23 @@ void SparsePoseGraph::WaitForAllComputations() {
   locker.Await([&notification]() { return notification; });
 }
 
+void SparsePoseGraph::FinishTrajectory(const int trajectory_id) {
+  AddWorkItem([this, trajectory_id]() REQUIRES(mutex_) {
+    CHECK_EQ(finished_trajectories_.count(trajectory_id), 0);
+    finished_trajectories_.insert(trajectory_id);
+
+    const auto& submap_data = optimization_problem_.submap_data();
```
Comment (ojura) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Can you make a pass through all submaps in `submap_data_` which match the given trajectory ID without accessing the `optimization_problem_`? Then you would not need to use the work queue at all.

Comment (jihoonl) [reply cartographer/mapping_2d/sparse_pose_graph.cc]: Submap states are updated within work queue thread always. See [ComputeConstraintsForScan](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L293), [AddSubmapFromProto](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L437), and [MarkSubmapAsTrimmed](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L670).

Comment (ojura) [reply cartographer/mapping_2d/sparse_pose_graph.cc]: Ah yes, good point. I actually posted [the same question](https://github.com/googlecartographer/cartographer/commit/7434e963697c2876add9658e1ff1af6399589215#commitcomment-24545844) to @wohe some time ago.

File: cartographer/mapping/sparse_pose_graph.h [hunk 2]
```diff
@@ -74,6 +74,9 @@ class SparsePoseGraph {
   SparsePoseGraph(const SparsePoseGraph&) = delete;
   SparsePoseGraph& operator=(const SparsePoseGraph&) = delete;
 
+  // Finish a trajectory
```
Comment (wohe) [inline cartographer/mapping/sparse_pose_graph.h]: `Finishes the given trajectory.`?

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 3]
```diff
@@ -365,6 +365,11 @@ void SparsePoseGraph::WaitForAllComputations() {
   locker.Await([&notification]() { return notification; });
 }
 
+void SparsePoseGraph::FinishTrajectory(const int trajectory_id) {
+  // TODO(jihoonl): Add a logic to notify trimmers to finish the given
+  // trajectory
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: nit: End sentence with a period.

File: cartographer/mapping_3d/sparse_pose_graph.cc [hunk 4]
```diff
@@ -386,6 +386,10 @@ void SparsePoseGraph::WaitForAllComputations() {
   locker.Await([&notification]() { return notification; });
 }
 
+void SparsePoseGraph::FinishTrajectory(const int trajectory_id) {
+  LOG(FATAL) << "Not yet implemented for 3D.";
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: This would break 3D localization which we now support. Change to a `TODO` comment similar to 2D?

**Pull request merged**